### PR TITLE
Support logit_bias in v1 Sampler

### PR DESCRIPTION
**Pull request by houseroad**
Introduce logit_bias support in v1 Sampler.

Tested with new test cases:
pytest tests/v1/sample/test_sampler.py -k "test_sampler_logit_bias"
pytest tests/v1/worker/test_gpu_input_batch.py

### CODE REVIEW

File: vllm/v1/sample/sampler.py [hunk 1]
```diff
@@ -166,3 +166,14 @@ def apply_penalties(
...
+
+    def apply_logits_bias(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+        for i, logit_bias in enumerate(sampling_metadata.logit_bias):
+            if logit_bias is not None:
+                for token_id, bias in logit_bias.items():
+                    logits[i, token_id] += bias
```
Comment (WoosukKwon) [inline vllm/v1/sample/sampler.py:178]: Can we add a comment like `TODO: This is extremely slow. Optimize this.`?

Comment (WoosukKwon) [reply vllm/v1/sample/sampler.py:178]: @njhill @robertgshaw2-redhat Although this implementation is a bit slow, I'm comfortable merging the PR since I haven't found a way to optimize it yet, and getting the functionality in is our top priority. What do you think?

**Review by WoosukKwon — COMMENTED** [review]
Thanks for the PR! Let some minor comments.

File: vllm/v1/sample/sampler.py (hunk 1 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py:178]: Yeah, I agree. We can write some customized kernel to handle such things in c++. Also we may change the representation of logit_bias from dict to key value pair.

I can create some TODO as follow up.

Comment (WoosukKwon) [reply vllm/v1/sample/sampler.py:178]: @houseroad Sg. Could you please update the PR?

File: vllm/v1/engine/async_llm.py
Comment (WoosukKwon) [inline vllm/v1/engine/async_llm.py:1]: QQ: What is this change for?

Comment (houseroad) [reply vllm/v1/engine/async_llm.py:1]: Sorry, this is due to some internal changes and our export flow.

For short term, I should just stick to the OSS workstream. :-)

File: vllm/v1/worker/gpu_input_batch.py [hunk 2]
```diff
@@ -463,6 +473,20 @@ def make_lora_inputs(
...
         return prompt_lora_mapping, token_lora_mapping, active_lora_requests
 
+    def _construct_logit_bias(logit_bias: Union[Dict[int, float], Dict[str, float]]) -> Dict[int, float]:
+        try:
+            # Convert token_id to integer
+            # Clamp the bias between -100 and 100 per OpenAI API spec
+            clamped_logit_bias: Dict[int, float] = {
+                int(token_id): min(100.0, max(-100.0, bias))
+                for token_id, bias in logit_bias.items()
+            }
```
[the remark this answers is not in the document]
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py:484]: So any suggestion on this?

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_input_batch.py:484]: @houseroad Thanks for the explanation. I think we should move it to `sampling_params.py`.

Also, I think we should avoid raising any error in model runner because it will kill the entire server (while if it's in sampling params, the invalid request will not affect the server).

File: vllm/v1/sample/sampler.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/sample/sampler.py:181]: @WoosukKwon I think that's fine since we need the functionality asap. But I think it should be simple to vectorize this without any custom kernel.

We just need to maintain in the batch three one-dim tensors of the same length: 
- all the bias values concatenated (b)
- corresponding request indices (s)
- corresponding token ids (t)

We only need to update these when any requests with logit bias are added or removed from the batch.

Then we can just do `logits[(s, t)] += v`

Comment (houseroad) [reply vllm/v1/sample/sampler.py:181]: Updated the PR. Please let me know if you want me directly jump to the optimized solution.

File: vllm/sampling_params.py [hunk 3]
```diff
@@ -244,7 +244,7 @@ def from_optional(
     ) -> "SamplingParams":
         if logit_bias is not None:
             logit_bias = {
-                int(token): bias
+                int(token): min(100.0, max(-100.0, bias))
```
Comment (WoosukKwon) [inline vllm/sampling_params.py:247]: nit: Could you add the original comment about why we clamp this value? 😅

File: vllm/v1/sample/sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/sampler.py:181]: @njhill @houseroad Considering that the code is pretty isolated, I think we can merge the PR first and have a followup PR to optimize it.

Comment (WoosukKwon) [reply vllm/v1/sample/sampler.py:181]: @njhill Re your idea: Each request's `logits_bias` has different lengths. How do we handle that (with the persistent batch)?

File: vllm/sampling_params.py (hunk 3 above)
Comment (houseroad) [reply vllm/sampling_params.py:249]: Sure, just added.

File: vllm/v1/sample/sampler.py (hunk 1 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py:181]: I think we can have ragged format representation, 3 tensors, tensor a for length/offset of each batch, tensor b for token ids, tensor c for bias. And one torch op takes these 3 tensors as inputs, and we leverage C++ logic to handle it. It should be much faster then the current logic.

Maybe in the SamplingMeta or param, we should just preproces things like this.

The additional overhead is that once we update the batch, we may need to generate new tensors, which should be acceptable.

**Review by WoosukKwon — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by WoosukKwon** [comment]
@houseroad Seems like we need to fix a unit test: https://buildkite.com/vllm/ci/builds/13363#0194fe70-1618-4cb6-ae2d-8380ee8a2041 😓
